xla/debug_options_flags.cc

```diff
@@ -222,6 +222,11 @@ DebugOptions DefaultDebugOptionsIgnoringFlags() {
 
   opts.set_xla_gpu_enable_dot_strength_reduction(true);
 
+  // Minimum combined size of matrices in matrix multiplication to
+  // be rewritten to cuBLAS or Triton kernel call.
+  const int64_t kDefaultMinGemmRewriteSize = 100;
```

Comment:
Why is that the right threshold value? Do we know that it's always better to use Triton/cuBLAS above this?

Not disagreeing, but I'd like documentation on why that is a good default (I'd also like to measure this over microbenchmarks & have this be part of the commit message to justify the change as well).

Reply:
It turns out the threshold is quite conservative, see below. This was on H100 - on V100 the results are even more pronounced (0.015s vs 0.006s  at 16x16x16 in favor of skipped-rewriting).

I have also tested bf16, but there the numbers seem to be similar (10x  regardless of rewriting, presumably because of some padding.

```
H100 (1M float32)
  
   8 (2x2x2     ): SkipRewrite: 0.00008   Rewritten: 0.00238           
  12 (2x2x4     ): SkipRewrite: 0.00012   Rewritten: 0.00230
  12 (4x2x2     ): SkipRewrite: 0.00012   Rewritten: 0.00235
  16 (2x4x2     ): SkipRewrite: 0.00019   Rewritten: 0.00235
  16 (4x2x4     ): SkipRewrite: 0.00028   Rewritten: 0.00229
  20 (2x2x8     ): SkipRewrite: 0.00024   Rewritten: 0.00230
  20 (8x2x2     ): SkipRewrite: 0.00022   Rewritten: 0.00232
  24 (2x4x4     ): SkipRewrite: 0.00023   Rewritten: 0.00225
  24 (4x4x2     ): SkipRewrite: 0.00024   Rewritten: 0.00231
  24 (4x2x8     ): SkipRewrite: 0.00035   Rewritten: 0.00230
  24 (8x2x4     ): SkipRewrite: 0.00030   Rewritten: 0.00236
  32 (2x8x2     ): SkipRewrite: 0.00028   Rewritten: 0.00226
  32 (4x4x4     ): SkipRewrite: 0.00025   Rewritten: 0.00232
  32 (8x2x8     ): SkipRewrite: 0.00044   Rewritten: 0.00231
  36 (2x2x16    ): SkipRewrite: 0.00034   Rewritten: 0.00230
  36 (16x2x2    ): SkipRewrite: 0.00034   Rewritten: 0.00189
  40 (2x4x8     ): SkipRewrite: 0.00035   Rewritten: 0.00232
  40 (4x2x16    ): SkipRewrite: 0.00046   Rewritten: 0.00232
  40 (8x4x2     ): SkipRewrite: 0.00028   Rewritten: 0.00234
  40 (16x2x4    ): SkipRewrite: 0.00047   Rewritten: 0.00234
  48 (2x8x4     ): SkipRewrite: 0.00032   Rewritten: 0.00225
  48 (4x8x2     ): SkipRewrite: 0.00030   Rewritten: 0.00236
  48 (4x4x8     ): SkipRewrite: 0.00039   Rewritten: 0.00236
  48 (8x4x4     ): SkipRewrite: 0.00035   Rewritten: 0.00232
  48 (8x2x16    ): SkipRewrite: 0.00075   Rewritten: 0.00233
  48 (16x2x8    ): SkipRewrite: 0.00075   Rewritten: 0.00235
  64 (2x16x2    ): SkipRewrite: 0.00032   Rewritten: 0.00291
  64 (4x8x4     ): SkipRewrite: 0.00040   Rewritten: 0.00231
  64 (8x4x8     ): SkipRewrite: 0.00051   Rewritten: 0.00231
  64 (16x2x16   ): SkipRewrite: 0.00138   Rewritten: 0.00240
  72 (2x4x16    ): SkipRewrite: 0.00045   Rewritten: 0.00236
  72 (16x4x2    ): SkipRewrite: 0.00043   Rewritten: 0.00231
  80 (2x8x8     ): SkipRewrite: 0.00041   Rewritten: 0.00230
  80 (4x4x16    ): SkipRewrite: 0.00056   Rewritten: 0.00228
  80 (8x8x2     ): SkipRewrite: 0.00040   Rewritten: 0.00228
  80 (16x4x4    ): SkipRewrite: 0.00056   Rewritten: 0.00230
  96 (2x16x4    ): SkipRewrite: 0.00042   Rewritten: 0.00295
  96 (4x16x2    ): SkipRewrite: 0.00046   Rewritten: 0.00287
  96 (4x8x8     ): SkipRewrite: 0.00052   Rewritten: 0.00234
  96 (8x8x4     ): SkipRewrite: 0.00050   Rewritten: 0.00229
  96 (8x4x16    ): SkipRewrite: 0.00084   Rewritten: 0.00234
  96 (16x4x8    ): SkipRewrite: 0.00082   Rewritten: 0.00237
============================================================
 128 (4x16x4    ): SkipRewrite: 0.00059   Rewritten: 0.00289
 128 (8x8x8     ): SkipRewrite: 0.00070   Rewritten: 0.00236
 128 (16x4x16   ): SkipRewrite: 0.00138   Rewritten: 0.00250
 144 (2x8x16    ): SkipRewrite: 0.00065   Rewritten: 0.00238
 144 (16x8x2    ): SkipRewrite: 0.00061   Rewritten: 0.00237
 160 (2x16x8    ): SkipRewrite: 0.00064   Rewritten: 0.00286
 160 (4x8x16    ): SkipRewrite: 0.00077   Rewritten: 0.00238
 160 (8x16x2    ): SkipRewrite: 0.00062   Rewritten: 0.00294
 160 (16x8x4    ): SkipRewrite: 0.00078   Rewritten: 0.00233
 192 (4x16x8    ): SkipRewrite: 0.00079   Rewritten: 0.00292
 192 (8x16x4    ): SkipRewrite: 0.00082   Rewritten: 0.00294
 192 (8x8x16    ): SkipRewrite: 0.00113   Rewritten: 0.00239
 192 (16x8x8    ): SkipRewrite: 0.00113   Rewritten: 0.00240
 256 (8x16x8    ): SkipRewrite: 0.00108   Rewritten: 0.00296
 256 (16x8x16   ): SkipRewrite: 0.00184   Rewritten: 0.00250
 288 (2x16x16   ): SkipRewrite: 0.00110   Rewritten: 0.00288
 288 (16x16x2   ): SkipRewrite: 0.00103   Rewritten: 0.00293
 320 (4x16x16   ): SkipRewrite: 0.00125   Rewritten: 0.00289
 320 (16x16x4   ): SkipRewrite: 0.00127   Rewritten: 0.00297
 384 (8x16x16   ): SkipRewrite: 0.00179   Rewritten: 0.00292
 384 (16x16x8   ): SkipRewrite: 0.00181   Rewritten: 0.00292
 512 (16x16x16  ): SkipRewrite: 0.00264   Rewritten: 0.00302
 ```

Reply:
For completeness, here are the bf16 numbers:

```
H100 (1M bf16 matmuls)
   8 (2x2x2     ): SkipRewrite: 0.00085   Rewritten: 0.00125
  12 (2x2x4     ): SkipRewrite: 0.00091   Rewritten: 0.00134
  12 (4x2x2     ): SkipRewrite: 0.00085   Rewritten: 0.00129
  16 (2x4x2     ): SkipRewrite: 0.00090   Rewritten: 0.00131
  16 (4x2x4     ): SkipRewrite: 0.00089   Rewritten: 0.00134
  20 (2x2x8     ): SkipRewrite: 0.00088   Rewritten: 0.00133
  20 (8x2x2     ): SkipRewrite: 0.00090   Rewritten: 0.00133
  24 (2x4x4     ): SkipRewrite: 0.00093   Rewritten: 0.00135
  24 (4x4x2     ): SkipRewrite: 0.00086   Rewritten: 0.00128
  24 (4x2x8     ): SkipRewrite: 0.00090   Rewritten: 0.00136
  24 (8x2x4     ): SkipRewrite: 0.00096   Rewritten: 0.00139
  32 (2x8x2     ): SkipRewrite: 0.00083   Rewritten: 0.00128
  32 (4x4x4     ): SkipRewrite: 0.00091   Rewritten: 0.00137
  32 (8x2x8     ): SkipRewrite: 0.00080   Rewritten: 0.00125
  36 (2x2x16    ): SkipRewrite: 0.00129   Rewritten: 0.00144
  36 (16x2x2    ): SkipRewrite: 0.00137   Rewritten: 0.00158
  40 (2x4x8     ): SkipRewrite: 0.00090   Rewritten: 0.00137
  40 (4x2x16    ): SkipRewrite: 0.00134   Rewritten: 0.00154
  40 (8x4x2     ): SkipRewrite: 0.00090   Rewritten: 0.00134
  40 (16x2x4    ): SkipRewrite: 0.00141   Rewritten: 0.00160
  48 (2x8x4     ): SkipRewrite: 0.00091   Rewritten: 0.00139
  48 (4x8x2     ): SkipRewrite: 0.00082   Rewritten: 0.00126
  48 (4x4x8     ): SkipRewrite: 0.00094   Rewritten: 0.00140
  48 (8x4x4     ): SkipRewrite: 0.00097   Rewritten: 0.00144
  48 (8x2x16    ): SkipRewrite: 0.00117   Rewritten: 0.00135
  48 (16x2x8    ): SkipRewrite: 0.00117   Rewritten: 0.00135
  64 (2x16x2    ): SkipRewrite: 0.00127   Rewritten: 0.00156
  64 (4x8x4     ): SkipRewrite: 0.00091   Rewritten: 0.00138
  64 (8x4x8     ): SkipRewrite: 0.00086   Rewritten: 0.00131
  64 (16x2x16   ): SkipRewrite: 0.00198   Rewritten: 0.00160
  72 (2x4x16    ): SkipRewrite: 0.00125   Rewritten: 0.00142
  72 (16x4x2    ): SkipRewrite: 0.00137   Rewritten: 0.00161
  80 (2x8x8     ): SkipRewrite: 0.00065   Rewritten: 0.00114
  80 (4x4x16    ): SkipRewrite: 0.00134   Rewritten: 0.00152
  80 (8x8x2     ): SkipRewrite: 0.00069   Rewritten: 0.00117
  80 (16x4x4    ): SkipRewrite: 0.00142   Rewritten: 0.00159
  96 (2x16x4    ): SkipRewrite: 0.00128   Rewritten: 0.00155
  96 (4x16x2    ): SkipRewrite: 0.00129   Rewritten: 0.00155
  96 (4x8x8     ): SkipRewrite: 0.00069   Rewritten: 0.00116
  96 (8x8x4     ): SkipRewrite: 0.00072   Rewritten: 0.00117
  96 (8x4x16    ): SkipRewrite: 0.00119   Rewritten: 0.00136
  96 (16x4x8    ): SkipRewrite: 0.00123   Rewritten: 0.00137
 128 (4x16x4    ): SkipRewrite: 0.00131   Rewritten: 0.00159
 128 (8x8x8     ): SkipRewrite: 0.00042   Rewritten: 0.00092
 128 (16x4x16   ): SkipRewrite: 0.00203   Rewritten: 0.00171
 144 (2x8x16    ): SkipRewrite: 0.00100   Rewritten: 0.00116
 144 (16x8x2    ): SkipRewrite: 0.00111   Rewritten: 0.00129
 160 (2x16x8    ): SkipRewrite: 0.00099   Rewritten: 0.00129
 160 (4x8x16    ): SkipRewrite: 0.00109   Rewritten: 0.00126
 160 (8x16x2    ): SkipRewrite: 0.00105   Rewritten: 0.00132
 160 (16x8x4    ): SkipRewrite: 0.00115   Rewritten: 0.00132
 192 (4x16x8    ): SkipRewrite: 0.00102   Rewritten: 0.00130
 192 (8x16x4    ): SkipRewrite: 0.00107   Rewritten: 0.00133
 192 (8x8x16    ): SkipRewrite: 0.00074   Rewritten: 0.00092
 192 (16x8x8    ): SkipRewrite: 0.00077   Rewritten: 0.00093
 256 (8x16x8    ): SkipRewrite: 0.00063   Rewritten: 0.00095
 256 (16x8x16   ): SkipRewrite: 0.00140   Rewritten: 0.00099
 288 (2x16x16   ): SkipRewrite: 0.00149   Rewritten: 0.00138
 288 (16x16x2   ): SkipRewrite: 0.00153   Rewritten: 0.00151
 320 (4x16x16   ): SkipRewrite: 0.00159   Rewritten: 0.00148
 320 (16x16x4   ): SkipRewrite: 0.00158   Rewritten: 0.00154
 384 (8x16x16   ): SkipRewrite: 0.00108   Rewritten: 0.00097
 384 (16x16x8   ): SkipRewrite: 0.00104   Rewritten: 0.00097
 512 (16x16x16  ): SkipRewrite: 0.00181   Rewritten: 0.00128
```

Reply:
Thank you so much for taking the time to benchmark this! Should we include these results in the commit message for traceability?

I also wonder what we could put in the code to document this threshold. Would it be reasonable to add something to the comment that says something like "This threshold is a conservative estimate and has been measured to be always beneficial (up to generally several times faster) on V100 and H100 GPUs. See openxla/xla #9319 for details."?